File: scripts/sync-dart-doc-comments.py
#!/usr/bin/env python3
from __future__ import annotations

import argparse
import difflib
import re
import sys
from dataclasses import dataclass
from pathlib import Path
from typing import Iterable, Tuple, Union


PROVIDER_START_RE = re.compile(r"<!--\s*\{@([A-Za-z0-9_]+)(?::[^}]*)?\}\s*-->")
PROVIDER_END_RE = re.compile(r"<!--\s*\{/([A-Za-z0-9_]+)\}\s*-->")
CONSUMER_START_RE = re.compile(
    r"^(?P<indent>\s*)///\s*<!--\s*\{=(?P<name>[A-Za-z0-9_]+)(?P<transforms>[^}]*)\}\s*-->\s*$"
)
CONSUMER_END_RE = re.compile(
    r"^(?P<indent>\s*)///\s*<!--\s*\{/(?P<name>[A-Za-z0-9_]+)\}\s*-->\s*$"
)
SUPPORTED_TRANSFORMS = {"replace", "trim", "trimStart", "trimEnd"}


@dataclass(frozen=True)
class ReplaceTransform:
    old: str
    new: str


Transform = Union[Tuple[str], Tuple[str, ReplaceTransform]]


@dataclass(frozen=True)
class ConsumerBlock:
    path: Path
    name: str
    start_index: int
    end_index: int
    transforms: tuple[Transform, ...]


class ParseError(RuntimeError):
    pass


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(
        description=(
            "Synchronize Dart /// doc comment consumers from markdown template "
            "provider blocks."
        )
    )
    parser.add_argument(
        "--check",
        action="store_true",
        help="Validate that Dart doc comment consumers are up to date.",
    )
    parser.add_argument(
        "--write",
        action="store_true",
        help="Rewrite stale Dart doc comment consumers (default).",
    )
    return parser.parse_args()


def decode_quoted(text: str, index: int) -> tuple[str, int]:
    if index >= len(text) or text[index] != '"':
        raise ParseError(f'Expected quoted string at offset {index}: {text!r}')

    index += 1
    value_chars: list[str] = []
    while index < len(text):
        char = text[index]
        if char == '\\':
            index += 1
            if index >= len(text):
                raise ParseError(f'Invalid trailing escape in transform: {text!r}')
            escape = text[index]
            escapes = {
                '"': '"',
                '\\': '\\',
                'n': '\n',
                'r': '\r',
                't': '\t',
            }
            value_chars.append(escapes.get(escape, escape))
            index += 1
            continue
        if char == '"':
            return ''.join(value_chars), index + 1
        value_chars.append(char)
        index += 1

    raise ParseError(f'Unterminated quoted string in transform: {text!r}')


def parse_transforms(raw: str) -> tuple[Transform, ...]:
    text = raw.strip()
    if not text:
        return ()

    index = 0
    transforms: list[Transform] = []
    while index < len(text):
        if text[index] != '|':
            raise ParseError(f'Expected `|` in transform list: {text!r}')
        index += 1

        name_start = index
        while index < len(text) and text[index] not in ':|':
            index += 1
        name = text[name_start:index].strip()
        if name not in SUPPORTED_TRANSFORMS:
            raise ParseError(f'Unsupported transform `{name}` in {text!r}')

        if name == 'replace':
            if index >= len(text) or text[index] != ':':
                raise ParseError(f'`replace` requires two quoted arguments: {text!r}')
            index += 1
            old, index = decode_quoted(text, index)
            if index >= len(text) or text[index] != ':':
                raise ParseError(f'`replace` requires two quoted arguments: {text!r}')
            index += 1
            new, index = decode_quoted(text, index)
            transforms.append((name, ReplaceTransform(old=old, new=new)))
        else:
            transforms.append((name,))

    return tuple(transforms)


def apply_transforms(content: str, transforms: Iterable[Transform]) -> str:
    rendered = content
    for transform in transforms:
        kind = transform[0]
        if kind == 'replace':
            replacement = transform[1]
            assert isinstance(replacement, ReplaceTransform)
            rendered = rendered.replace(replacement.old, replacement.new)
        elif kind == 'trim':
            rendered = rendered.strip()
        elif kind == 'trimStart':
            rendered = rendered.lstrip()
        elif kind == 'trimEnd':
            rendered = rendered.rstrip()
        else:
            raise ParseError(f'Unsupported transform {kind!r}')
    return rendered


def parse_providers(repo_root: Path) -> dict[str, str]:
    providers: dict[str, str] = {}
    for template_path in sorted(repo_root.glob('*.t.md')):
        lines = template_path.read_text().splitlines()
        index = 0
        while index < len(lines):
            start_match = PROVIDER_START_RE.search(lines[index])
            if not start_match:
                index += 1
                continue

            name = start_match.group(1)
            end_index = index + 1
            while end_index < len(lines):
                end_match = PROVIDER_END_RE.search(lines[end_index])
                if end_match and end_match.group(1) == name:
                    break
                end_index += 1
            else:
                raise ParseError(
                    f'Missing provider end tag for `{name}` in {template_path}'
                )

            if name in providers:
                raise ParseError(
                    f'Duplicate provider `{name}` found in {template_path} '
                    f'and another template file.'
                )
            providers[name] = '\n'.join(lines[index + 1 : end_index])
            index = end_index + 1
    return providers


def iter_consumer_blocks(path: Path) -> list[ConsumerBlock]:
    lines = path.read_text().splitlines()
    blocks: list[ConsumerBlock] = []
    index = 0
    while index < len(lines):
        start_match = CONSUMER_START_RE.match(lines[index])
        if not start_match:
            index += 1
            continue

        name = start_match.group('name')
        transforms = parse_transforms(start_match.group('transforms'))
        end_index = index + 1
        while end_index < len(lines):
            end_match = CONSUMER_END_RE.match(lines[end_index])
            if end_match and end_match.group('name') == name:
                break
            end_index += 1
        else:
            raise ParseError(f'Missing consumer end tag for `{name}` in {path}')

        blocks.append(
            ConsumerBlock(
                path=path,
                name=name,
                start_index=index,
                end_index=end_index,
                transforms=transforms,
            )
        )
        index = end_index + 1
    return blocks


def render_doc_comment(content: str) -> list[str]:
    lines = content.split('\n')
    if lines == ['']:
        return ['///']

    rendered: list[str] = []
    for line in lines:
        if line:
            rendered.append(f'/// {line}')
        else:
            rendered.append('///')
    return rendered


def update_file(path: Path, providers: dict[str, str], write: bool) -> tuple[bool, str]:
    original_text = path.read_text()
    lines = original_text.splitlines()
    blocks = iter_consumer_blocks(path)
    if not blocks:
        return False, ''

    updated_lines = list(lines)
    delta = 0
    changed = False

    for block in blocks:
        if block.name not in providers:
            raise ParseError(f'No provider found for consumer `{block.name}` in {path}')

        rendered = apply_transforms(providers[block.name], block.transforms)
        replacement = render_doc_comment(rendered)

        start = block.start_index + 1 + delta
        end = block.end_index + delta
        current = updated_lines[start:end]
        if current != replacement:
            updated_lines[start:end] = replacement
            delta += len(replacement) - len(current)
            changed = True

    if not changed:
        return False, ''

    new_text = '\n'.join(updated_lines)
    if original_text.endswith('\n'):
        new_text += '\n'

    diff = ''.join(
        difflib.unified_diff(
            original_text.splitlines(keepends=True),
            new_text.splitlines(keepends=True),
            fromfile=str(path),
            tofile=str(path),
        )
    )

    if write:
        path.write_text(new_text)

    return True, diff


def main() -> int:
    args = parse_args()
    write = not args.check or args.write

    repo_root = Path(__file__).resolve().parent.parent
    providers = parse_providers(repo_root)
    dart_files = sorted(repo_root.glob('packages/*/lib/**/*.dart'))

    changed_paths: list[Path] = []
    diffs: list[str] = []
    consumer_count = 0

    for path in dart_files:
        blocks = iter_consumer_blocks(path)
        if not blocks:
            continue
        consumer_count += len(blocks)
        changed, diff = update_file(path, providers, write=write)
        if changed:
            changed_paths.append(path)
            diffs.append(diff)

    mode = 'Updated' if write else 'Checked'
    if not changed_paths:
        print(
            f'{mode} {consumer_count} Dart doc comment consumer block(s); '
            'all are up to date.'
        )
        return 0

    if write:
        print(
            f'Updated {len(changed_paths)} file(s) across {consumer_count} '
            'Dart doc comment consumer block(s).'
        )
        return 0

    print(
        f'{len(changed_paths)} file(s) have stale Dart doc comment consumers '
        f'across {consumer_count} block(s).',
        file=sys.stderr,
    )
    for diff in diffs:
        print(diff, file=sys.stderr, end='')
    return 1


if __name__ == '__main__':
    try:
        raise SystemExit(main())
    except ParseError as error:
        print(f'error: {error}', file=sys.stderr)
        raise SystemExit(2)

File: scripts/sync_dart_doc_comments.dart
// ignore_for_file: parameter_assignments

import 'dart:io';

final _providerStartPattern = RegExp(
  r'<!--\s*\{@([A-Za-z0-9_]+)(?::[^}]*)?\}\s*-->',
);
final _providerEndPattern = RegExp(r'<!--\s*\{/([A-Za-z0-9_]+)\}\s*-->');
final _consumerStartPattern = RegExp(
  r'^(?<indent>\s*)///\s*<!--\s*\{=(?<name>[A-Za-z0-9_]+)(?<transforms>[^}]*)\}\s*-->\s*$',
);
final _consumerEndPattern = RegExp(
  r'^(?<indent>\s*)///\s*<!--\s*\{/(?<name>[A-Za-z0-9_]+)\}\s*-->\s*$',
);
const _supportedTransforms = {'replace', 'trim', 'trimStart', 'trimEnd'};

void main(List<String> args) {
  try {
    final check = args.contains('--check');
    final write = !check || args.contains('--write');
    final providers = _parseProviders(Directory.current);
    final dartFiles = _dartLibraryFiles();

    final changedPaths = <String>[];
    var consumerCount = 0;
    for (final file in dartFiles) {
      final blocks = _consumerBlocks(file);
      if (blocks.isEmpty) continue;
      consumerCount += blocks.length;
      final changed = _updateFile(file, providers, write: write);
      if (changed) changedPaths.add(file.path);
    }

    final mode = write ? 'Updated' : 'Checked';
    if (changedPaths.isEmpty) {
      stdout.writeln(
        '$mode $consumerCount Dart doc comment consumer block(s); all are up to date.',
      );
      return;
    }

    if (write) {
      stdout.writeln(
        'Updated ${changedPaths.length} file(s) across $consumerCount Dart doc comment consumer block(s).',
      );
      return;
    }

    stderr.writeln(
      '${changedPaths.length} file(s) have stale Dart doc comment consumers across $consumerCount block(s).',
    );
    for (final path in changedPaths) {
      stderr.writeln('- $path');
    }
    exitCode = 1;
  } on _ParseError catch (error) {
    stderr.writeln('error: ${error.message}');
    exitCode = 2;
  }
}

Map<String, String> _parseProviders(Directory repoRoot) {
  final providers = <String, String>{};
  final templates =
      repoRoot
          .listSync()
          .whereType<File>()
          .where((file) => file.path.endsWith('.t.md'))
          .toList()
        ..sort((a, b) => a.path.compareTo(b.path));

  for (final template in templates) {
    final lines = template.readAsLinesSync();
    var index = 0;
    while (index < lines.length) {
      final start = _providerStartPattern.firstMatch(lines[index]);
      if (start == null) {
        index += 1;
        continue;
      }

      final name = start.group(1)!;
      var endIndex = index + 1;
      while (endIndex < lines.length) {
        final end = _providerEndPattern.firstMatch(lines[endIndex]);
        if (end != null && end.group(1) == name) break;
        endIndex += 1;
      }
      if (endIndex >= lines.length) {
        throw _ParseError(
          'Missing provider end tag for `$name` in ${template.path}',
        );
      }
      if (providers.containsKey(name)) {
        throw _ParseError(
          'Duplicate provider `$name` found in ${template.path} and another template file.',
        );
      }
      providers[name] = lines.sublist(index + 1, endIndex).join('\n');
      index = endIndex + 1;
    }
  }
  return providers;
}

List<File> _dartLibraryFiles() {
  final packages = Directory('packages');
  if (!packages.existsSync()) return [];
  final files =
      packages
          .listSync(recursive: true)
          .whereType<File>()
          .where(
            (file) =>
                file.path.contains('/lib/') && file.path.endsWith('.dart'),
          )
          .toList()
        ..sort((a, b) => a.path.compareTo(b.path));
  return files;
}

List<_ConsumerBlock> _consumerBlocks(File file) {
  final lines = file.readAsLinesSync();
  final blocks = <_ConsumerBlock>[];
  var index = 0;
  while (index < lines.length) {
    final start = _consumerStartPattern.firstMatch(lines[index]);
    if (start == null) {
      index += 1;
      continue;
    }

    final name = start.namedGroup('name')!;
    final transforms = _parseTransforms(start.namedGroup('transforms')!);
    var endIndex = index + 1;
    while (endIndex < lines.length) {
      final end = _consumerEndPattern.firstMatch(lines[endIndex]);
      if (end != null && end.namedGroup('name') == name) break;
      endIndex += 1;
    }
    if (endIndex >= lines.length) {
      throw _ParseError('Missing consumer end tag for `$name` in ${file.path}');
    }

    blocks.add(_ConsumerBlock(name, index, endIndex, transforms));
    index = endIndex + 1;
  }
  return blocks;
}

bool _updateFile(
  File file,
  Map<String, String> providers, {
  required bool write,
}) {
  final originalText = file.readAsStringSync();
  final lines = originalText.split('\n');
  if (lines.isNotEmpty && lines.last == '') lines.removeLast();
  final blocks = _consumerBlocks(file);
  if (blocks.isEmpty) return false;

  final updated = [...lines];
  var delta = 0;
  var changed = false;
  for (final block in blocks) {
    final provider = providers[block.name];
    if (provider == null) {
      throw _ParseError(
        'No provider found for consumer `${block.name}` in ${file.path}',
      );
    }

    final replacement = _renderDocComment(
      _applyTransforms(provider, block.transforms),
    );
    final start = block.startIndex + 1 + delta;
    final end = block.endIndex + delta;
    final current = updated.sublist(start, end);
    if (!_listEquals(current, replacement)) {
      updated.replaceRange(start, end, replacement);
      delta += replacement.length - current.length;
      changed = true;
    }
  }

  if (!changed) return false;
  var newText = updated.join('\n');
  if (originalText.endsWith('\n')) newText += '\n';
  if (write) file.writeAsStringSync(newText);
  return true;
}

List<String> _renderDocComment(String content) {
  final lines = content.split('\n');
  if (lines.length == 1 && lines.single.isEmpty) return ['///'];
  return [for (final line in lines) line.isEmpty ? '///' : '/// $line'];
}

String _applyTransforms(String content, List<_Transform> transforms) {
  var rendered = content;
  for (final transform in transforms) {
    switch (transform.name) {
      case 'replace':
        rendered = rendered.replaceAll(
          transform.oldValue!,
          transform.newValue!,
        );
      case 'trim':
        rendered = rendered.trim();
      case 'trimStart':
        rendered = rendered.trimLeft();
      case 'trimEnd':
        rendered = rendered.trimRight();
      default:
        throw _ParseError('Unsupported transform `${transform.name}`');
    }
  }
  return rendered;
}

List<_Transform> _parseTransforms(String raw) {
  final text = raw.trim();
  if (text.isEmpty) return [];
  final transforms = <_Transform>[];
  var index = 0;
  while (index < text.length) {
    if (text[index] != '|') {
      throw _ParseError('Expected `|` in transform list: $text');
    }
    index += 1;

    final nameStart = index;
    while (index < text.length && text[index] != ':' && text[index] != '|') {
      index += 1;
    }
    final name = text.substring(nameStart, index).trim();
    if (!_supportedTransforms.contains(name)) {
      throw _ParseError('Unsupported transform `$name` in $text');
    }

    if (name == 'replace') {
      if (index >= text.length || text[index] != ':') {
        throw _ParseError('`replace` requires two quoted arguments: $text');
      }
      index += 1;
      final oldDecoded = _decodeQuoted(text, index);
      index = oldDecoded.$2;
      if (index >= text.length || text[index] != ':') {
        throw _ParseError('`replace` requires two quoted arguments: $text');
      }
      index += 1;
      final newDecoded = _decodeQuoted(text, index);
      index = newDecoded.$2;
      transforms.add(_Transform(name, oldDecoded.$1, newDecoded.$1));
    } else {
      transforms.add(_Transform(name));
    }
  }
  return transforms;
}

(String, int) _decodeQuoted(String text, int index) {
  if (index >= text.length || text[index] != '"') {
    throw _ParseError('Expected quoted string at offset $index: $text');
  }
  index += 1;
  final buffer = StringBuffer();
  while (index < text.length) {
    final char = text[index];
    if (char == r'\') {
      index += 1;
      if (index >= text.length) {
        throw _ParseError('Invalid trailing escape in transform: $text');
      }
      final escape = text[index];
      buffer.write(switch (escape) {
        '"' => '"',
        r'\' => r'\',
        'n' => '\n',
        'r' => '\r',
        't' => '\t',
        _ => escape,
      });
      index += 1;
      continue;
    }
    if (char == '"') return (buffer.toString(), index + 1);
    buffer.write(char);
    index += 1;
  }
  throw _ParseError('Unterminated quoted string in transform: $text');
}

bool _listEquals(List<String> a, List<String> b) {
  if (a.length != b.length) return false;
  for (var i = 0; i < a.length; i += 1) {
    if (a[i] != b[i]) return false;
  }
  return true;
}

final class _ConsumerBlock {
  const _ConsumerBlock(
    this.name,
    this.startIndex,
    this.endIndex,
    this.transforms,
  );
  final String name;
  final int startIndex;
  final int endIndex;
  final List<_Transform> transforms;
}

final class _Transform {
  const _Transform(this.name, [this.oldValue, this.newValue]);
  final String name;
  final String? oldValue;
  final String? newValue;
}

final class _ParseError implements Exception {
  const _ParseError(this.message);
  final String message;
}
